File: traversal/src/test/scala/overflowdb/traversal/testdomains/gratefuldead/Artist.scala
package overflowdb.traversal.testdomains.gratefuldead

import overflowdb.traversal.Traversal
import overflowdb.{Graph, NodeFactory, NodeLayoutInformation, NodeRef, PropertyKey}

import scala.jdk.CollectionConverters._

class Artist(graph: Graph, id: Long) extends NodeRef[ArtistDb](graph, id) {
  override def label: String = Artist.Label

  def name: String = get.name

  /* Artist <-- sungBy --- Song */
  def sangSongs: Traversal[Song] = get.sangSongs

  /* Artist <-- writtenBy --- Song */
  def wroteSongs: Traversal[Song] = get.wroteSongs
}

object Artist {
  val Label = "artist"
  val LabelId = 4

  object Properties {
    val Name = new PropertyKey[String](PropertyNames.Name)
  }

  object PropertyNames {
    val Name = "name"
    val all: Set[String] = Set(Name)
  }

  val factory: NodeFactory[ArtistDb] = new NodeFactory[ArtistDb]() {
    override def forLabel: String = Label
    override def forLabelId() = LabelId
    override def createNode(ref: NodeRef[ArtistDb]) = new ArtistDb(ref)
    override def createNodeRef(graph: Graph, id: Long) = new Artist(graph, id)
    override val layoutInformation = Artist.layoutInformation
  }

  val layoutInformation: NodeLayoutInformation =
    new NodeLayoutInformation(
      LabelId,
      PropertyNames.all.asJava,
      Nil.asJava,
      List(SungBy.layoutInformation, WrittenBy.layoutInformation).asJava
    )
}

File: core/src/test/java/overflowdb/testdomains/gratefuldead/Artist.java
package overflowdb.testdomains.gratefuldead;

import overflowdb.NodeFactory;
import overflowdb.NodeRef;
import overflowdb.Graph;

public class Artist extends NodeRef<ArtistDb> {
  public static final String label = "artist";
  public static final String NAME = "name";

  public Artist(Graph graph, long id) {
    super(graph, id);
  }

  public String name() {
    return get().name();
  }

  @Override
  public String label() {
    return Artist.label;
  }

  public static NodeFactory<ArtistDb> factory = new NodeFactory<ArtistDb>() {
    @Override
    public String forLabel() {
      return Artist.label;
    }

    @Override
    public int forLabelId() {
      return ArtistDb.layoutInformation.labelId;
    }

    @Override
    public ArtistDb createNode(NodeRef<ArtistDb> ref) {
      return new ArtistDb(ref);
    }

    @Override
    public Artist createNodeRef(Graph graph, long id) {
      return new Artist(graph, id);
    }
  };
}
